timeseries/src/autogluon/timeseries/predictor.py

```diff
@@ -187,57 +186,109 @@ def __init__(
                 )
-        if self.ignore_time_index:
-            df = df.get_reindexed_view(freq="S")
+            return data
+        else:
+            raise TypeError(
+                f"{name} must be a TimeSeriesDataFrame or pandas.DataFrame or string (path to data) "
+                f"but received an object of type {type(data)}."
+            )
+
+    def _check_and_prepare_data_frame(
+        self,
+        data: Union[TimeSeriesDataFrame, pd.DataFrame, str],
+        name: str = "data",
+    ) -> TimeSeriesDataFrame:
+        """Ensure that TimeSeriesDataFrame has a sorted index, valid frequency, and contains no missing values.
+
+        If self.freq is None, then self.freq of the predictor will be set to the frequency of the data.
+
+        Parameters
+        ----------
+        data : Union[TimeSeriesDataFrame, pd.DataFrame, str]
+            Data as a data frame or path to file storing the data.
+        name : str
+            Name of the data that will be used in log messages (e.g., 'train_data', 'tuning_data', or 'data').
+
+        Returns
+        -------
+        df : TimeSeriesDataFrame
+            Preprocessed data in TimeSeriesDataFrame format.
+        """
+        df = self._to_data_frame(data, name=name)
         # MultiIndex.is_monotonic_increasing checks if index is sorted by ["item_id", "timestamp"]
         if not df.index.is_monotonic_increasing:
             df = df.sort_index()
             df._cached_freq = None  # in case frequency was incorrectly cached as IRREGULAR_TIME_INDEX_FREQSTR
-        if df.freq is None:
-            raise ValueError(
-                "Frequency not provided and cannot be inferred. This is often due to the "
-                "time index of the data being irregularly sampled. Please ensure that the "
-                "data set used has a uniform time index, or create the `TimeSeriesPredictor` "
-                "setting `ignore_time_index=True`."
-            )
-        # Check if frequency is supported
+
+        # Ensure that data has a regular frequency that matches the predictor frequency
+        if self.freq is None:
+            if df.freq is None:
+                raise ValueError(
+                    f"Frequency of {name} is not provided and cannot be inferred. Please set the expected data "
+                    f"frequency when creating the predictor with `TimeSeriesPredictor(freq=...)` or ensure that "
+                    f"the data has a regular time index with `{name}.to_regular_index(freq=...)`"
+                )
+            else:
+                self.freq = df.freq
+                logger.info(f"Inferred data frequency: {df.freq}")
+        else:
+            if df.freq != self.freq:
+                logger.warning(f"{name} with frequency '{df.freq}' has been resampled to frequency '{self.freq}'.")
+                df = df.convert_frequency(freq=self.freq)
```

Comment:
@tonyhoo Suppose user creates `TimeSeriesPredictor(freq="D")` and passes hourly data to it. Which of the two alternatives is more reasonable?
1. Automatically resample hourly data to the daily frequency + notify the user
2. Raise an exception saying that provided data is incompatible with daily frequency